flask/app.py

```diff
@@ -1370,6 +1392,32 @@ def url_defaults(self, f):
         self.url_default_functions.setdefault(None, []).append(f)
         return f
 
+    def _find_error_handler(self, e):
+        """Finds a registered error handler for the request’s blueprint.
+        If neither blueprint nor App has a suitable handler registered, returns None
+        """
+        exc_class, code = self._ensure_exc_class(e)
+        
+        def find_superclass(d):
+            if not d:
+                return None
+            for superclass in exc_class.mro():
+                if superclass is BaseException:
+                    return None
+                handler = d.get(superclass)
+                if handler is not None:
+                    return handler
```

Comment:
you mean `d[exc_class]`. `e` is an exception instance.

and that would create an edge case:

```python
app = Flask(__name__)

# hierarchy: Exception → OSError → ConnectionError → ConnectionAbortedError

@app.errorhandler(OSError)
def handle_os_error(e):
    ...

app.run()

somewhere during runtime some connection aborts, a ConnectionAbortedError instance is raised. now, due to your caching: `d[ConnectionAbortedError] is d[OSError]`

also during runtime (after that), some new error handler is registered. (not inconceivable):

```python
@app.errorhandler(ConnectionError)
def handle_connection_error(e):
    ...
```

well fuck, now another ConnectionAbortedError instance being raised won’t trigger the newly registered error handler but the cached one!

---

we can circumvent this by instead having a separate dict holding all cached handlers, and clearing all superclasses of a newly registered exception from that cache handler dict. like so:

```python
# in __init__:
self._error_handler_cache = {}

# in _register_error_handler:
for superclass in exc_class.mro():
    try:
        del self._error_handler_cache[superclass]
    except KeyError:
        pass

# in _find_error_handler → find_superclass (after “if not d: return None”):
handler = self._error_handler_cache.get(exc_class)
if handler is not None:
    return handler

# in _find_error_handler → find_superclass (before returning the handler):
self._error_handler_cache[exc_class] = handler
```